Empire/AmpAdsInjector.php

```diff
@@ -0,0 +1,102 @@
+<?php
+
+namespace Empire;
+
+class AmpAdsInjector extends \AMP_Base_Sanitizer {
+    public function sanitize() {
+        $ampConfig = $this->args['ampConfig'];
+        $adsConfig = $this->args['adsConfig'];
+        $getTargeting = $this->args['getTargeting'];
+
+        if ($this->checkAdsBlocked($adsConfig['adRules'], $getTargeting())) {
+            return;
+        }
+
+        foreach ($ampConfig['forPlacement'] as $key => $amp) {
+            $place = $adsConfig['forPlacement'][$key];
+            $this->injectAds($amp['component'], $place['selectors'], $place['limit']);
+        }
+    }
+
+    public function injectAds($ad, $selectors, $limit) {
+        $count = 0;
+        $transformer = \FluentDOM::getXPathTransformer();
+        foreach ($selectors as $selector) {
+            $path = $transformer->toXpath($selector);
+
+            foreach ($this->dom->xpath->query($path) as $elem) {
+                $component = $this->nodeFromHtml($ad);
+                # TODO: respect relative option
+                $elem->insertBefore($component);
+                $count++;
+
+                if ($count == $limit) {
+                    return;
+                }
+            }
+        }
+    }
+
+    public function nodeFromHtml($html) {
+        $fragment = $this->dom::fromHtmlFragment($html);
+        $exportBody = $fragment->getElementsByTagName('body')->item(0);
+
+        $importFragment = $this->dom->createDocumentFragment();
+        while ($exportBody->firstChild) {
+            $importNode = $exportBody->removeChild($exportBody->firstChild);
+            $importNode = $this->dom->importNode($importNode, true);
+            $importFragment->appendChild($importNode);
+        }
+        return $importFragment;
+    }
+
+    public function checkAdsBlocked($adRules, $targeting) {
+        foreach ($adRules as $rule) {
+            if (!$rule['enabled']) {
+                continue;
+            }
+
+            $url = $targeting['url'];
+            $components = [];
+            switch ($rule['component']) {
+                case 'PATH':
+                    $components = [parse_url($url)['path'] ?? '/'];
+                    break;
+                case 'URL':
```

Comment:
We don't expose the `URL` component in the dashboard:

![image](https://user-images.githubusercontent.com/4452484/129964385-8d1c4f61-23c6-400a-98f9-5a0616d6bcfc.png)

Is this something that's programmatically set and returned by the API? If we aren't using it, we should probably remove it.